// react-simple-img/lib/logic/observerStart.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultConfig = undefined;
exports.onIntersection = onIntersection;
exports.default = observerStart;

var _imageLoader = require('../utils/imageLoader');

var _imageLoader2 = _interopRequireDefault(_imageLoader);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultConfig = exports.defaultConfig = {
  rootMargin: '0px 0px',
  threshold: [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
};
function onIntersection(entries) {
  for (var i = 0, len = entries.length; i < len; i++) {
    var _entries$i = entries[i],
        _intersectionRatio = _entries$i.intersectionRatio,
        _target = _entries$i.target;

    if (_intersectionRatio > 0) {
      _imageLoader2.default.call(this, _target);
    }
  }
}

function observerStart() {
  var _this = this;

  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : defaultConfig;

  if (!window.IntersectionObserver) require('intersection-observer');
  // $FlowIgnoreLine:
  var observer = new IntersectionObserver(function (entries) {
    return onIntersection.call(_this, entries);
  }, config);
  if (this) return observer;

  window.__REACT_SIMPLE_IMG__ = {
    observer: observer,
    imgLoadingRefs: new Map()
  };

  return undefined;
}
