// react-simple-img/lib/simpleImg.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleImg = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactSimpleAnimate = require('react-simple-animate');

var _validImgSrc = require('./utils/validImgSrc');

var _validImgSrc2 = _interopRequireDefault(_validImgSrc);

var _simpleImgProvider = require('./simpleImgProvider');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var commonStyle = {
  position: 'absolute',
  top: 0,
  left: 0,
  height: '100%'
};
var rootStyle = {
  position: 'relative',
  overflow: 'hidden',
  display: 'flex'
};
var defaultDisappearStyle = { opacity: 0 };
var defaultImgPlaceholder = 'data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=';
var defaultPlaceholderColor = 'white';
var onCompleteStyle = { display: 'none' };
var fullWidthStyle = { width: '100%' };
var hiddenStyle = { visibility: 'hidden' };

var SimpleImg = exports.SimpleImg = function (_React$PureComponent) {
  _inherits(SimpleImg, _React$PureComponent);

  function SimpleImg() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, SimpleImg);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = SimpleImg.__proto__ || Object.getPrototypeOf(SimpleImg)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      loaded: false,
      isDocumentLoad: false
    }, _this.setDocumentLoaded = function () {
      _this.setState({
        isDocumentLoad: true
      });
    }, _this.element = _react2.default.createRef(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(SimpleImg, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (window.__REACT_SIMPLE_IMG__ && document.readyState === 'complete') {
        window.__REACT_SIMPLE_IMG__.observer.observe(this.element.current);
      } else if (document.readyState === 'complete') {
        this.setDocumentLoaded();
      } else {
        window.addEventListener('load', this.setDocumentLoaded);
      }
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps, prevState) {
      var _this2 = this;

      var _props = this.props,
          appendImageRef = _props.appendImageRef,
          useContext = _props.useContext,
          removeImageRef = _props.removeImageRef,
          mountedImages = _props.mountedImages,
          removeImgLoadingRef = _props.removeImgLoadingRef,
          isContextDocumentLoad = _props.isContextDocumentLoad;

      var element = this.element.current;

      if (useContext) {
        if ((!prevProps.isContextDocumentLoad && isContextDocumentLoad || !prevState.isDocumentLoad && this.state.isDocumentLoad) && element) {
          appendImageRef(element);
          removeImgLoadingRef(element);
        }

        if (element && mountedImages.has(element)) {
          setTimeout(function () {
            return _this2.setState({
              loaded: true
            });
          });
          removeImageRef(element);
        }
      } else if (!prevState.isDocumentLoad && this.state.isDocumentLoad) {
        window.__REACT_SIMPLE_IMG__.observer.observe(element);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('load', this.setDocumentLoaded);
      if (!this.element.current) return;
      var _props2 = this.props,
          removeImgLoadingRef = _props2.removeImgLoadingRef,
          removeImageRef = _props2.removeImageRef,
          useContext = _props2.useContext;

      var element = this.element.current;

      if (useContext && element) {
        removeImgLoadingRef(element);
        removeImageRef(element);
      } else {
        if (!window.__REACT_SIMPLE_IMG__) return;

        var _window$__REACT_SIMPL = window.__REACT_SIMPLE_IMG__,
            observer = _window$__REACT_SIMPL.observer,
            imgLoadingRefs = _window$__REACT_SIMPL.imgLoadingRefs;

        observer.unobserve(element);

        if (imgLoadingRefs.has(element)) {
          imgLoadingRefs.get(element).src = '';
          imgLoadingRefs.delete(element);
        }
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props3 = this.props,
          src = _props3.src,
          className = _props3.imgClassName,
          wrapperClassName = _props3.wrapperClassName,
          width = _props3.width,
          height = _props3.height,
          alt = _props3.alt,
          srcSet = _props3.srcSet,
          sizes = _props3.sizes,
          animationDuration = _props3.animationDuration,
          _props3$animationEndS = _props3.animationEndStyle,
          animationEndStyle = _props3$animationEndS === undefined ? defaultDisappearStyle : _props3$animationEndS,
          _props3$placeholder = _props3.placeholder,
          placeholder = _props3$placeholder === undefined ? defaultPlaceholderColor : _props3$placeholder;
      var loaded = this.state.loaded;

      var isValidImgSrc = (0, _validImgSrc2.default)(placeholder);
      var inlineStyle = _extends({}, commonStyle, !isValidImgSrc ? { background: placeholder } : null, {
        height: height
      });
      var imgPlaceholder = isValidImgSrc ? placeholder : defaultImgPlaceholder;
      var isSrcSetFulfilled = this.element.current && this.element.current.src !== imgPlaceholder;

      return _react2.default.createElement(
        'span',
        { style: rootStyle, className: wrapperClassName },
        _react2.default.createElement('img', _extends({ width: width, height: height, sizes: sizes, className: className }, {
          alt: alt,
          ref: this.element,
          src: loaded ? src : imgPlaceholder,
          srcSet: loaded ? srcSet : '',
          'data-src': src,
          'data-srcset': srcSet,
          style: _extends({}, !isValidImgSrc && !loaded && !isSrcSetFulfilled ? hiddenStyle : null)
        })),
        isValidImgSrc && _react2.default.createElement(_reactSimpleAnimate.Animate, _extends({
          play: loaded,
          durationSeconds: animationDuration,
          endStyle: _extends({}, inlineStyle, animationEndStyle, !isValidImgSrc ? fullWidthStyle : null),
          onCompleteStyle: onCompleteStyle
        }, !isValidImgSrc ? {
          startStyle: _extends({}, inlineStyle, fullWidthStyle)
        } : null, {
          render: function render(_ref2) {
            var style = _ref2.style;
            return _react2.default.createElement('img', _extends({ width: width, height: height, className: className }, { style: _extends({}, inlineStyle, style), alt: alt, src: placeholder }));
          }
        }))
      );
    }
  }]);

  return SimpleImg;
}(_react2.default.PureComponent);

SimpleImg.defaultProps = {
  animationDuration: 0.3
};

exports.default = function (props) {
  return _react2.default.createElement(
    _simpleImgProvider.SimpleImgContext.Consumer,
    null,
    function (values) {
      return _react2.default.createElement(SimpleImg, _extends({}, props, values));
    }
  );
};
